src/rules/noUnnecessaryElseRule.ts

```diff
@@ -0,0 +1,98 @@
+/**
+ * @license
+ * Copyright 2018 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        description: Lint.Utils.dedent`
+            Disallows else block when the If block contains control flow statements, such as \`return\`, \`continue\`,
+            \`break\` or \`throw\`.`,
+        descriptionDetails: "",
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        rationale: Lint.Utils.dedent`
+            When control flow statements,
+            such as \`return\`, \`continue\`, \`break\` or \`throw\` is written inside \`if\` block,
+            Then \`else\` block becomes unnecessary.`,
+        ruleName: "no-unnecessary-else",
+        type: "functionality",
+        typescriptOnly: false,
+    };
+    /* tslint:disable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(name: string): string {
+        return `If block contains \`${name}\` statement. Consider replacing the contents of else block outside of the block.`;
+    }
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk);
+    }
+}
+
+type JumpStatement =
+    | ts.BreakStatement
+    | ts.ContinueStatement
+    | ts.ThrowStatement
+    | ts.ReturnStatement;
+
+function walk(ctx: Lint.WalkContext<void>): void {
+    let inElse = false;
```

Comment:
It looks like `walk` will consider any jump statement within an `if` block to make that block not need an `else`. What if there's one within a nested `if` statement?

```ts
if (something) {
    if (anotherThing) {
        return;
    } else {
        // Unnecessary else block
    }
} else {
    // Reasonable else block
}
```

Instead of walking down within the `if`, I think you can get away with just checking whether its _last_ statement is a jump statement. In other words, I think you should be able to just use `if (isIfStatement(node)) {` and check the last statement within the `thenStatement` and the node's `elseStatement`.

Reply:
Done https://github.com/palantir/tslint/pull/4502/commits/89c814eb7648cbe716af317adb05add694305430